request-example.go

```diff
@@ -213,6 +213,12 @@ func (fs *root) Filelist(r *Request) (ListerAt, error) {
 		}
 		return listerat(list), nil
 	case "Stat":
+		if file.symlink != "" {
+			file, err = fs.fetch(file.symlink)
```

Comment:
This functionality should be in a separate `"Lstat"` method. The two are different system calls for a reason. And `“Stat”` is the one that does not follow symlinks.

Reply:
`stat` should follow the symlink, it was broken before, please try this sample code:

```
package main

import (
	"fmt"
	"io/ioutil"
	"os"
)

func main() {
	fileName := "testfile.txt"
	fileNameLink := "testfile.link"
	os.Remove(fileName)
	os.Remove(fileNameLink)
	err := ioutil.WriteFile(fileName, []byte("hello"), os.ModePerm)
	if err != nil {
		fmt.Printf("error creating file: %v\n", err)
		os.Exit(1)
	}
	err = os.Symlink(fileName, fileNameLink)
	if err != nil {
		fmt.Printf("error creating symlink: %v\n", err)
		os.Exit(1)
	}
	fi, err := os.Stat(fileName)
	if err != nil {
		fmt.Printf("stat error: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("stat file info: %+v size: %v is symlink: %v\n", fi, fi.Size(), fi.Mode()&os.ModeSymlink)

	fi, err = os.Stat(fileNameLink)
	if err != nil {
		fmt.Printf("stat error: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("stat link info: %+v size: %v is symlink: %v\n", fi, fi.Size(), fi.Mode()&os.ModeSymlink)

	fi, err = os.Lstat(fileNameLink)
	if err != nil {
		fmt.Printf("lstat error: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("lstat link info: %+v size: %v is symlink: %v\n", fi, fi.Size(), fi.Mode()&os.ModeSymlink)
}
```

Reply:
No, I misread the text from `man stat`:
```
lstat() is identical to stat(), except that if path is a symbolic link, then the link itself is stat-ed, not the file that it refers to.
```

🤦‍♀️

Reply:
I think we still need an `Lstat` Request handler, though?

Reply:
We should be able to do it with a `fallthrough` as well:
```
case "Stat":
  if file.symlink != "" {
    file, err := fs.fetch(file.symlink)
    if err != nil { … }
  }
  fallthrough
case "Lstat":
  return listerat([]os.FileInfo{file}), nil
```

Reply:
The point here is to keep backward compatibility. If we add a `Lstat` method to `Filelist` in request-example.go then all the implementors need to do the same. I added an optional interface so `Lstat` is called only if the optional LstatFileLister interface is implemented otherwise Lstat will become a Stat as before. I agree that this is not ideal

Reply:
But this isn’t `FileList`, it’s a `Request` handler on `switch r.Method { … }`. I don’t see how expanding the functionality of a struct in `request-example.go` would require all implementors to do the same. 😕 Especially, since no where in this `Filelist` receiver method, we are calling a `.Lstat` on anything.

Reply:
Maybe I'm missing something, are we talking about the change I did to the Filelister interface implementation for the in memory [request example](https://github.com/drakkan/sftp/blob/lstat/request-example.go#L185)?

Reply:
Yes, I’m talking about this specific piece of code that this thread is attached to, which is in `request-example.go`. Good that we now have the correct behavior for `"Stat"` Method, but we should also include an `"Lstat"` example as well, right?

Reply:
`Filelist` is an interface called by the request server and implemented by other software. We are talking about the sample implementation inside pkg/sftp (request-example.go), [here](https://github.com/drakkan/sftpgo/blob/master/sftpd/handler.go#L193) is the implementation for this interface in SFTPGo.

If the request server set the request method to `Lstat` then both request-example.go and the SFTPGo implementation must be updated to handle the new `Lstat` method. The current request-server go never set the request method to `Lstat`. It sets the method to `Stat` for both `stat` and `lstat` so the implementors don't expect the `Lstat` method. For this reson I wrote a backward compatible path. To handle `lstat` request-example must implements the `LstatFileLister` and handle `lstat` inside this optional interface. In SFTPGo I'll do the same. If the interface is not implemented lstat will be hidden as with the current code and the `Filelister` interface will receive `Stat` as request method.

If you prefer I can update the patch to be backward incompatible

Reply:
OK, I think I see where I’m getting absolutely confused. If the handler implements `LstatFileLister`, then we call the `Lstat(…)` receiver, otherwise we call `Filelist(…)` with `r.Method = "Stat"`?

So, we’re handling two things in very different methods. (Which I suppose is how we did the `OpenFile` extension?)

I think I have been so confused, because I didn’t expect us to handle things in very different directions. 🤔 hm…

Reply:
> OK, I think I see where I’m getting absolutely confused. If the handler implements `LstatFileLister`, then we call the `Lstat(…)` receiver, otherwise we call `Filelist(…)` with `r.Method = "Stat"`?
> 

yes, this is to keep backward compatibility if we call `Filelist(…)` with `r.Method = "Lstat"` all the current implementors must be updated

> So, we’re handling two things in very different methods. (Which I suppose is how we did the `OpenFile` extension?)
> 
> I think I have been so confused, because I didn’t expect us to handle things in very different directions.  hm…

a backward incompatible change, calling `Filelist(…)` with `r.Method = "Lstat"`, is cleaner but we could cause troubles